File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/circular_array_rotation.ts
// https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

function main() {
  const [_, nRotation, nQueries]: number[] = readNumbers();
  const arr: number[] = readNumbers();
  for (const q of queries(arr, nRotation, nQueries)) console.log(q);
}

function readNumbers(): number[] {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// T: O(nQueries)
// S: O(1) extra space
function* queries(
  arr: number[],
  nRotation: number,
  nQueries: number,
): Generator<number | undefined> {
  const offset: number = nRotation % arr.length;
  for (let i = 0; i < nQueries; i++) {
    const query = +readLine();
    yield arr.at(query - offset);
  }
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}

File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/circular_array_rotation.py
# https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

def main():
    _, n_rotation, n_queries = read_numbers()
    arr = read_numbers()
    print(*queries(arr, n_rotation, n_queries), sep='\n')


def read_numbers():
    return list(map(int, input().split()))


# T: O(n_queries)
# S: O(1) extra space
def queries(arr, n_rotation, n_queries):
    offset = n_rotation % len(arr)
    for _ in range(n_queries):
        query = int(input())
        yield arr[query - offset]


if __name__ == '__main__':
    main()
